Add custom parameters to authorize and logout endpoint

eva-mueller-coremedia opened the pull request:
This PR
* provides a different approach to solve https://github.com/jenkinsci/oic-auth-plugin/issues/229 resp. https://github.com/jenkinsci/oic-auth-plugin/pull/241
* solves https://github.com/jenkinsci/oic-auth-plugin/issues/443

Additionally, some housekeeping (fixing typos, simply statements) has been done. 

The main change has been done in eb0d2ba1ed13aa1fd318fe3ad96cd876d49c7efb:
* Add new input fields for login resp. logout query parameters in the Jenkins management backend
* `maybeOpenIdLogoutEndpoint` has been refactored when it comes to combining all parameters
* Login query parameters are not allowed to overwrite keys as defined in `org.pac4j.oidc.config.OidcConfiguration`
* Logout query parameters are not allowed to overwrite keys as set in `maybeOpenIdLogoutEndpoint` like `id_token_hint`, `state`, `post_logout_redirect_uri`

### Testing done

This change has been tested by unit tests as well as local testing agains AWS Cognito.

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md in your own repository 
-->

jtnord reviewed (commented):
Hi @eva-mueller-coremedia thanks for this.

I have had a quick look and have a few comments around the UX side.

The extra parameters would IIUC not be needed in the majority of conformant IDPs, as such do you thing they should be hidden behind an [advanced section](https://reports.jenkins.io/core-taglib/jelly-taglib-ref.html#form:advanced)?

the use of a single string for multiple keyvalue pairs (and the encoding /splitting shenaningans) would seem to be error prone (and it lacks good warnings on invalid input).  (a value may need to encode `&` and then you need to escape it)
Would making the KeyValue a Describable and using a repeatable option using `repeatableProperty` or the like (which can then perform individual valuation of the key and the value).  NB not sur how this would look in CasC.

eva-mueller-coremedia commented:
@jtnord Thanks for the feedback, will try to follow a key value pair pattern.

If you don't mind, I would like to remove the UI at all - only provide it to be set via CasC resp. Java/Groovy code. Any objections?

jtnord commented:
> @jtnord Thanks for the feedback, will try to follow a key value pair pattern.
> 
> If you don't mind, I would like to remove the UI at all - only provide it to be set via CasC resp. Java/Groovy code. Any objections?

it's generally considered bad to have something exposed in the UI and not CasC and vice versa.  Not exposing in the UI can lead to it getting overridden if someone saves the Jenkins form in the UI (which I say you should never do as if you manage by code you should have a read only system UI, but it is useful on test systems where you want to make changes in the UI and then export the config for applying to a different server).

eva-mueller-coremedia commented:
> > @jtnord Thanks for the feedback, will try to follow a key value pair pattern.
> > If you don't mind, I would like to remove the UI at all - only provide it to be set via CasC resp. Java/Groovy code. Any objections?
> 
> it's generally considered bad to have something exposed in the UI and not CasC and vice versa. Not exposing in the UI can lead to it getting overridden if someone saves the Jenkins form in the UI (which I say you should never do as if you manage by code you should have a read only system UI, but it is useful on test systems where you want to make changes in the UI and then export the config for applying to a different server).

Understood. 👍  I will have a look to implement it the way you described initially. Thanks for the explanation!

eva-mueller-coremedia commented on src/main/java/org/jenkinsci/plugins/oic/OicServerWellKnownConfiguration.java:
```diff
@@ -219,12 +219,12 @@ public FormValidation doCheckWellKnownOpenIDConfigurationUrl(
 
         @POST
-        public FormValidation doCheckOverrideScopes(@QueryParameter String overrideScopes) {
+        public FormValidation doCheckScopesOverride(@QueryParameter String scopesOverride) {
```
@jtnord The form validation for scopes did not work since the jelly uses [`scopesOverride`](https://github.com/jenkinsci/oic-auth-plugin/blob/master/src/main/resources/org/jenkinsci/plugins/oic/OicServerWellKnownConfiguration/config.jelly#L10) as field name. 

Fixed this with this PR or should I open a dedicated PR?

eva-mueller-coremedia replied:
I will open a dedicated PR for this.

eva-mueller-coremedia commented:
@jtnord Thanks for the review. Will take care after my 1 week vacation.

eva-mueller-coremedia commented on src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:
```diff
@@ -505,7 +533,7 @@ ProxyAwareResourceRetriever getResourceRetriever() {
     }
 
-    private OidcConfiguration buildOidcConfiguration() {
+    private OidcConfiguration buildOidcConfiguration(boolean addCustomLoginParams) {
```
I thought of this as kind of "optimization" but looking at this some days later and with your review comment, I will dissolve this.

eva-mueller-coremedia commented on src/main/java/org/jenkinsci/plugins/oic/OicQueryParameterConfiguration.java:
```diff
@@ -0,0 +1,84 @@
+
+    public String getQueryParamValueDecoded() {
+        return paramValue != null
+                ? URLEncoder.encode(paramValue, StandardCharsets.UTF_8).trim()
```
Oh, thanks for pointing to this. Will fix the method name

eva-mueller-coremedia commented on src/main/java/org/jenkinsci/plugins/oic/OicQueryParameterConfiguration.java:
```diff
@@ -0,0 +1,84 @@
+        return paramValue;
+    }
+
+    public String getQueryParamNameDecoded() {
+        return paramName != null
+                ? URLEncoder.encode(paramName, StandardCharsets.UTF_8).trim()
+                : null;
+    }
```
Oh, thanks for pointing to this. Will fix the method name

eva-mueller-coremedia commented on src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:
```diff
@@ -357,6 +364,9 @@ protected Object readResolve() throws ObjectStreamException {
         // ensure escapeHatchSecret is encrypted
         this.setEscapeHatchSecret(this.escapeHatchSecret);
 
+        this.setLoginQueryParamNameValuePairs(this.loginQueryParamNameValuePairs);
+        this.setLogoutQueryParamNameValuePairs(this.logoutQueryParamNameValuePairs);
```
Yes, seem to be leftover. Will remove the calls

eva-mueller-coremedia commented:
@jtnord How to re-trigger the failed Jenkins run - there has been a `java.net.SocketTimeoutException: Read timed out` when downloading artifacts from azure...

jtnord commented:
> @jtnord How to re-trigger the failed Jenkins run - there has been a `java.net.SocketTimeoutException: Read timed out` when downloading artifacts from azure...

For non maintainers, adding a new commit.

eva-mueller-coremedia commented:
@jtnord The PR is now ready for a re-review.

eva-mueller-coremedia commented:
Sorry for nudging again - @jtnord @michael-doubez would be great if you could have a look.

jtnord reviewed (commented):
The code looks reasonable but some things look a little weird and I am wondering if thebUX can be improved at all.
I wanted to try and get some time to see if this can be improved and submit a PR to the PR as doing so after merging would possibly require even more data migration that we already have.realistically this is not going to happen this week.

eva-mueller-coremedia commented:
> realistically this is not going to happen this week.

Thanks for this estimate. Would be great to see this feature in the (near) future since AWS Cognito requires some additional query parameters.

It would be great if you could keep me/subscribers posted about the state of this PR from time to time. Thanks in advance.

jtnord commented:
@eva-mueller-coremedia I filed 
https://github.com/eva-mueller-coremedia/oic-auth-plugin/pull/1 against your PR.

Moves the parameters to their own types and added validation for them so it's validated in the UI.  Also ensures the config can never be invalid.
Tidies up the jelly slightly.

jtnord commented on src/main/java/org/jenkinsci/plugins/oic/LoginQueryParameter.java:
```diff
@@ -0,0 +1,38 @@
+    public static class DescriptorImpl extends AbstractKeyValueDescribable.DescriptorImpl<LoginQueryParameter> {
+
+        @Override
+        public FormValidation doCheckKey(@QueryParameter String key) {
```
There is nothing sensitive here.
Whilst we could add a permission check we are only doing string comparisons

eva-mueller-coremedia replied:
@jtnord Unfortunately, I do not know how to resolve all the problems found by the github-advanced-security. Can you help me here or step in? That would be great.

Also there is already one dismissed problem. Dunno who or how this got dismissed.

jtnord replied:
I can take a look.
I dismissed one as a false positive. (`key` was not sensitive in that context)

eva-mueller-coremedia commented:
> @eva-mueller-coremedia I filed [eva-mueller-coremedia#1](https://github.com/eva-mueller-coremedia/oic-auth-plugin/pull/1) against your PR.
> 
> Moves the parameters to their own types and added validation for them so it's validated in the UI. Also ensures the config can never be invalid. Tidies up the jelly slightly.

Thank you for the PR. I removed the commented out test.

eva-mueller-coremedia commented on src/main/java/org/jenkinsci/plugins/oic/AbstractKeyValueDescribable.java:
```diff
@@ -0,0 +1,88 @@
+            throw new Descriptor.FormException(valueValidation.getMessage(), "value");
+        }
+        this.key = key.trim();
+        this.value = value == null ? "" : value.trim();
```
Hi @jtnord - I added some more test cases and encountered for LogoutQueryParameters a NPE when the value is null.

Is this solution OK for you?

eva-mueller-coremedia commented:
@jtnord @michael-doubez @basil - May I politely ask if there is a change that this PR get's reviewed resp. merged soon or are you busy with other things right now? There for nudging again...
